_posts/2024-04-20-coroutines.md

```diff
@@ -0,0 +1,348 @@
+But there is no explicit support for coroutines.
+
+By its nature, all variables in the coroutine that survive suspension points form a state.
+And yet that state is kinda problematic for us, as other parts of the code can't access it while it is suspended.
+(For inspection/logging/data recording...)
+
+In this scenario, I am afraid that it might take a while until our debug tools are smart enough to give an easy life with coroutines in the system. Mostly: what if something went wrong in the coroutine? how do you inspect the state the coroutine is in?
+
+### Suggestions
+
+We talked about what coroutines are, and what are potential problems, now the question is: What can we use them for?
+
+For that, I am afraid it is hard to give a good answer.
+The functionality is simply quite new and I suspect that it will take a while (years) until we gain enough experience to be sure in which ways it is beneficial.
+
+That said, based on my (little) experience with the usage of coroutines in embedded, I have some suggestions and patterns that might be interesting.
+Mostly to pique the interest of the community in these, as I believe the potential is here.
+
+#### IO Coroutine
+
+
+I think that the IO example above is the most obvious way we can benefit from coroutines embedded.
+That is, we can use coroutines as an abstraction to interact with peripherals that do IO operations.
+
+To describe this, I will use a personal experience with `i2c_coroutine`.
+We can design it as follows:
+
+When `i2c_coroutine` is used as the type of the coroutine, the developer can interact with the `i2c` peripheral by `suspending` the coroutine with an `awaiter` that initiates operation on the `i2c` peripheral, the coroutine is resumed once that operation finishes.
+That is, we use the suspension point to allow interaction with the peripheral.
+
+For example:
+
+```cpp
+i2c_coroutine interact_with_device(uint8_t addr)
+{
+    std::array<uint8_t, 42> some_data;
+    co_await i2c_transmit_awaiter(addr, some_data);
+
+    std::span<const uint8_t> data = co_await i2c_receive_awaiter(addr, 42);
+    // do something with data
+}
+```
+
+This has the benefit that while the `i2c operation` is processed by the peripheral, the main loop can be busy doing something else, and the execution is not stuck in the `interact_with_device` with busy waiting.
+And we do not have to pay for that by having complexity (manual state machine doing the interaction) or having threads (which bring in other problems).
+
+Given the bus nature of `i2c`, it is also quite easy to achieve sharing of the `i2c bus` with multiple `device drivers` for various devices on the bus itself.
+We can just implement the `i2c_coroutine round_robin_run(std::span<i2c_coroutine> coros)` coroutine that uses round robin to share access to the peripheral between multiple coroutines (devices).
```

Comment:
I think this section would be very interesting to elaborate. Other strategies for handling concurrency for this scenario, and how you would implement them.

Reply:
I added some round robin implementation, wil lthink about others